src/openloco/windows/construction/ConstructionTab.cpp

```diff
@@ -0,0 +1,1276 @@
+                call(0x00460781, regs3);
+
+                if (regs3.ax == -32768)
+                    return;
+
+                mapPos.x = regs3.ax;
+                mapPos.y = regs3.bx;
+
+                getConstructionHeight(mapPos, &roadHeight, true);
+
+                _byte_113605D = 0;
+            }
+            input::cancel_tool();
+
+            auto ebx = 0;
+            if (input::has_key_modifier(input::key_modifier::shift) || _byte_113605D != 1)
+            {
+                auto roadPiece = common::roadPieces[_byte_1136065];
+                i = 0;
+                auto height = 0;
+
+                while (roadPiece[i].index != 0xFF)
+                {
+                    if (height > roadPiece[i].z)
+                        height = roadPiece[i].z;
+                    i++;
+                }
+
+                roadHeight -= height;
+                roadHeight -= 16;
+                ebx = 2;
+
+                if (input::has_key_modifier(input::key_modifier::shift))
+                {
+                    ebx = 0x80000000;
+                    roadHeight -= 16;
+                }
+            }
+            else
+            {
+                ebx = 1;
+                roadHeight = _word_1135FFE;
+            }
+
+            while (true)
+            {
+                _constructionHover = 0;
+                _byte_113607E = 0;
+                _x = mapPos.x;
+                _y = mapPos.y;
+                _word_1135FB8 = roadHeight;
+                _byte_522096 = 0;
+                _byte_1136066 = 0;
+
+                common::activateSelectedConstructionWidgets();
+                auto window = WindowManager::find(WindowType::construction);
+
+                if (window == nullptr)
+                    return;
+
+                _byte_508F09 = _byte_508F09 | (1 << 0);
+
+                on_mouse_up(window, widx::construct);
+
+                _byte_508F09 = _byte_508F09 & ~(1 << 0);
+
+                if (_dword_1135F42 == 0x80000000)
+                {
+                    if (gGameCommandErrorText != string_ids::error_can_only_build_above_ground)
+                    {
+                        ebx--;
+                        if (ebx != 0)
+                        {
+                            roadHeight -= 16;
+                            if (roadHeight >= 0)
+                            {
+                                if (ebx < 0)
+                                {
+                                    continue;
+                                }
+                                else
+                                {
+                                    roadHeight += 32;
+                                    continue;
+                                }
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    _byte_113607E = 1;
```

Comment:
This branch is to close the error message windows if the placement was succesful on the nth try. The 133607E is probably to prevent the audio playing on error.

Reply:
113607E is only tested at `0x4A07E2` if it is 1 it ors edi with a thing. `edi = lastSelectedMods << 16 | word_1135FB8`. So far the only function that uses edi after calling `getTrackPieceId` is `construction__draw`, but it immediately shifts it right by 16 leaving just `lastSelectedMods`. The  `getTrackPieceId`  function doesn't include 0x4A07E2, should I put a note stating this?

Reply:
leave as is was just comparing it to OpenRCT2 code.